site/_posts/2021-11-23-announcing-fusionauth-1-31.md

```diff
@@ -0,0 +1,49 @@
+---
+layout: blog-post
+title: Announcing FusionAuth 1.31
+description: This release includes a number of bug fixes as well as enhancements to reconcile lambda functionality.
+author: Dan Moore
+image: blogs/release-1-31/product-update-fusionauth-1-31.png
+category: blog
+tags: topic-troubleshooting feature-advanced-threat-detection
+excerpt_separator: "<!--more-->"
+---
+
+We're excited to announce the release of version 1.31.0 of FusionAuth. This version shipped on November 18, 2021. 1.31.0 includes a ton of bug fixes and improvements to the Identity Providers feature.
+
+<!--more-->
+
+This release contained a number of features, enhancements, and bug fixes. Please see the [release notes](/docs/v1/tech/release-notes/#version-1-31-0) for a full breakdown of the changes between 1.30 and 1.31. 
+
+There are a few items worth calling out.
+
+## Modify the username and email in a reconcile lambda
+
+FusionAuth lambdas are JavaScript functions that run at specific points in the authentication or authorization process. Identity Providers are external sources of account data, such as Facebook, Google, SAML or OIDC servers. Identity Provider reconcile lambdas run when an Identity Provider returns after a succesful login. 
+
+Previous to this release, with a few exceptions, the username and email claims could not be modified in a reconcile lambda. This increases flexibility when you have a remote identity datasource which does not provide either an email or a username. Sometimes identity linking can suffice, but there are other times when it does not. However, any time code can modify login Ids, make sure you threat model out the ramifications, particularly the danger of inadvertent account takeover.
+
```

Comment:
Might want to add a note about this introducing a possibly unexpected flow. If they are modify the configured linking data (change username if you are linking on username, then we will call the reconcile lambda a second time it a user matches that new linking data. This allows the reconcile a chance to act on the now linked user object.

If you don't modify that linking data, or no matching user is found using that modified data, then we don't bother calling the reconcile a second time.

Also, you can't modify the non-linking strategy property (i.e. - if you are linking on email then modifying username will be ignored because that could cause collisions if you were to change the linking strategy later)